NIFI-7406: PutAzureCosmosRecord Processor

sjyang18 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Functionally it is equivalent to PutMongoRecord Processor in nifi-mongodb-bundle, but this processor will use the cosmos native sql API. It is record-oriented Put Processor that uses the cosmos sql api.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

alopresto commented:
I updated the PR title as I believe it was linked to the wrong Jira (NIFI-7407 is not related to this work).

alopresto commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,336 @@
+            .build();
+
+    static final PropertyDescriptor CONNECTION_SERVICE = new PropertyDescriptor.Builder()
+        .name("azure-cosmos-db-connection-service")
```
Thanks for including clear `name` and `displayName` values for each of these property descriptors.

sjyang18 replied:
Thanks for reviewing & fixing my mistake. I also added displayName to "charactor-set" property.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml:
```diff
@@ -65,21 +87,74 @@
+            </exclusions>
+        </dependency>
+        <dependency>
+                <groupId>com.azure</groupId>
+                <artifactId>azure-cosmos</artifactId>
+                <version>4.0.1-beta.2</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>com.azure</groupId>
+                        <artifactId>azure-core</artifactId>
+                    </exclusion>
+                    <exclusion>
+                        <groupId>com.google.guava</groupId>
+                        <artifactId>guava</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
```
Looks like this is indented too much. Should be 4 spaces.

sjyang18 replied:
fixed

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+    static final String CONSISTENCY_CONSISTENT_PREFIX = "CONSISTENT_PREFIX";
+    static final String CONSISTENCY_EVENTUAL = "EVENTUAL";
+
+    static final Relationship REL_SUCCESS = new Relationship.Builder().name("success")
+    .description("All FlowFiles that are written to Azure Cosmos DB are routed to this relationship").build();
+
+    static final Relationship REL_FAILURE = new Relationship.Builder().name("failure")
+    .description("All FlowFiles that cannot be written to Azure Cosmos DB are routed to this relationship").build();
```
Can you style these like on L65.

sjyang18 replied:
fixed

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+        .displayName("Azure Cosmos DB URI")
+        .description("Azure Cosmos DB URI, typically of the form: https://{databaseaccount}.documents.azure.com:443/. Note this host URL is for Azure Cosmos DB with CORE SQL API")
+        .required(false)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Would it make sense to use `URL_VALIDATOR` or `URI_VALIDATOR` here?

sjyang18 replied:
fixed and moved to AzureCosmosDBUtils class.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+    static final PropertyDescriptor DB_ACCESS_KEY = new PropertyDescriptor.Builder()
+        .name("azure-cosmos-db-key")
+        .displayName("Azure Cosmos DB Access Key")
+        .description("Azure Cosmos DB Access Key from Azure Portal (Settings->Keys)")
```
I do like the concept of pointing to where to get the info. Do you have a sense of how stable that will be though? If it makes sense here I think it would end up making sense for the other properties too.

sjyang18 replied:
I tried to give the general statement so that user may figure out if there will be changes in Azure portal.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    static final PropertyDescriptor CONSISTENCY = new PropertyDescriptor.Builder()
```
This is a general comment for the properties. We're a lot more verbose with, e.g., the storage processors or the Kafka processors. What do the consistency levels mean here?

sjyang18 replied:
added more descriptions and moved to AzureCosmosDBUtils class for reuse across classes.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    static final PropertyDescriptor PARTITION_KEY = new PropertyDescriptor.Builder()
```
Similar comment, how does this impact usage?

sjyang18 replied:
done

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+
+    protected CosmosClient cosmosClient;
+    protected CosmosContainer container;
+    protected AzureCosmosDBConnectionService connectionService = null;
```
This will automatically default to null.

sjyang18 replied:
done

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+        getCosmosDocumentContainer(context);
+    }
+
+    protected void createCosmosClient(final String uri, final String accessKey, final ConsistencyLevel clevel) throws NullPointerException {
+        try {
+            ConnectionPolicy connectionPolicy = ConnectionPolicy.getDefaultPolicy();
+            this.cosmosClient = new CosmosClientBuilder()
+                .endpoint(uri)
+                .key(accessKey)
+                .connectionPolicy(connectionPolicy)
+                .consistencyLevel(clevel)
+                .buildClient();
+        } catch (NullPointerException e) {
+            getLogger().error("Failed to build cosmosClient due to {}",  e);
+            throw e;
+        }
+    }
```
What's the scenario where you get a `NullPointerException` here? It seems like something we should handle. Also you don't have to add a `throws NullPointerException`.

sjyang18 replied:
I had try catch from the implementation with previous SDK. Now we don't need to catch NullPointerException. I dropped the try and catch completely.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AbstractAzureCosmosDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+            new CosmosContainerProperties(containerID, "/"+partitionKey);
+
+        //  Create container with 1000 RU/s by default if Not exists
+        this.container = database.createContainerIfNotExists(containerProperties, 1000).getContainer();
```
Should this 1000 value be a property?

MuazmaZ replied:
This is the default at the Azure portal, so it seems reasonable to use.

sjyang18 replied:
I agree with Muazma. The right place to control Cosmos RU is from Azure portal. My code is is intended to create a container if not exists. This RU can be updated thru Azure portal and cosmos control panel at any time, and which will be mis-matched if we provide the property to this NIFI processor.

MuazmaZ commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,186 @@
+            try {
+                this.container.createItem(record);
+            }catch (ConflictException e) {
+                //  insert with unique id is expeected, but incase test data contains data with duplicate id(s).
```
typo 'expected'

sjyang18 replied:
fixed

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,186 @@
+            error = true;
+        } finally {
+            if (!error) {
+                session.getProvenanceReporter().send(flowFile, getURI(context));
+                session.transfer(flowFile, REL_SUCCESS);
```
Is there a reason this just isn't inside the `try() {}`?

sjyang18 replied:
It seems that session.transfer should be called outside try catch block. When I moved them inside try and catch block, I am getting the following exception. I will keep those two lines in finally block as PutMongoRecord does, not inside try catch:


[ERROR] testOnTriggerWithFlatRecords(org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest)  Time elapsed: 0.256 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: FlowFile[0,850021980779208.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed
        at org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest.testOnTriggerWithFlatRecords(PutAzureCosmosDBRecordTest.java:128)
Caused by: java.lang.IllegalStateException: FlowFile[0,850021980779208.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed

[ERROR] testOnTriggerWithNestedRecords(org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest)  Time elapsed: 0.013 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: FlowFile[0,850022330204874.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed
        at org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest.testOnTriggerWithNestedRecords(PutAzureCosmosDBRecordTest.java:190)
Caused by: java.lang.IllegalStateException: FlowFile[0,850022330204874.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,186 @@
+            }
+        } catch (SchemaNotFoundException | IOException | MalformedRecordException | CosmosClientException e) {
+            logger.error("PutAzureCosmoDBRecord failed with error:", e);
+            session.transfer(flowFile, REL_FAILURE);
```
There are two classes of errors here: (1) `SchemaNotFoundException` and `MalformedRecordException`, which we should rethrow as a `ProcessException` and (2) `CosmosClientException` which we should add a `context.yield();` for.

sjyang18 replied:
changed as you suggested.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/cosmos/document/AzureCosmosDBConnectionControllerService.java:
```diff
@@ -0,0 +1,188 @@
+                " and provides access to that connection to other AzureCosmosDB-related components."
+)
+public class AzureCosmosDBConnectionControllerService extends AbstractControllerService implements AzureCosmosDBConnectionService {
+    static final String CONSISTENCY_STRONG = "STRONG";
+    static final String CONSISTENCY_BOUNDED_STALENESS= "BOUNDED_STALENESS";
+    static final String CONSISTENCY_SESSION = "SESSION";
+    static final String CONSISTENCY_CONSISTENT_PREFIX = "CONSISTENT_PREFIX";
+    static final String CONSISTENCY_EVENTUAL = "EVENTUAL";
```
We have these constants in two places. Maybe pull out into another class or enum or something.

sjyang18 replied:
moved this definition to AzureCosmosDBUtils class and reused.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/cosmos/document/AzureCosmosDBConnectionControllerService.java:
```diff
@@ -0,0 +1,188 @@
+    @OnStopped
+    public final void closeClient() {
+        if (this.cosmosClient != null) {
+            cosmosClient.close();
```
Does this ever throw any exceptions? If so then I think it should get wrapped in a try-catch ignore so `cosmosClient = null` definitely happens.

sjyang18 replied:
done.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITAbstractAzureCosmosDBDocument.java:
```diff
@@ -0,0 +1,198 @@
+
+    private static final Properties CONFIG;
+
+    private static final String CREDENTIALS_FILE = System.getProperty("user.home") + "/azure-credentials.PROPERTIES";
```
Can we just assume the `cosmosURI` and `cosmosKey` for the IT are being passed in on the command line or a properties file specified on it?

sjyang18 replied:
I think this is an IT assumption in nifi-azure-bundle. I followed the same assumption from the storage IT (AbstractAzureStorageIT.java). Did I mis-understand your suggestion?

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITAbstractAzureCosmosDBDocument.java:
```diff
@@ -0,0 +1,198 @@
+    @AfterClass
+    public static void dropTestDBAndContainer() throws CosmosClientException {
+        resetTestCosmosConnection();
+        if(container != null) {
+            container.delete();
+            container = null;
+        }
+        if(cdb != null) {
+            cdb.delete();
+            cdb = null;
+
+        }
+        if(client != null){
+            client.close();
+            client = null;
+        }
```
Similar comment on try-catch ignore.

sjyang18 replied:
added try catch blocs

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITAbstractAzureCosmosDBDocument.java:
```diff
@@ -0,0 +1,198 @@
+            "select * from c order by c._ts", queryOptions, JsonNode.class );
+
+        response.forEach(data ->{
+            System.out.println(data);
```
Change to log message.

sjyang18 replied:
dropped the println

sjyang18 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml:
```diff
@@ -65,21 +87,74 @@
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-core</artifactId>
-            <version>2.10.3</version>
+            <version>${jackson.version}</version>
```
I override this jackson.version variable value to 2.10.3 in nifi-auzure-bundle/pom.xml.  So, it is the same value.

sjyang18 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml:
```diff
@@ -56,6 +56,28 @@
+        </dependency>
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-identity</artifactId>
```
dropped. It came from managed identity PR and this dependency is not needed for this PR.

sjyang18 commented:
@jfrazee I have rebased, resolved merge conflict, and updated with Cosmos SDK release version. I did forced-push since github rejected my update with the following error 'Updates were rejected because the tip of your current branch is behind its remote counterpart'. Will this work?

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/MockTestBase.java:
```diff
@@ -0,0 +1,74 @@
+            testRunner.setProperty(AbstractAzureCosmosDBProcessor.DATABASE_NAME, MOCK_DB_NAME);
+            testRunner.setProperty(AbstractAzureCosmosDBProcessor.CONTAINER_ID,MOCK_CONTAINER_ID);
+            testRunner.setProperty(AbstractAzureCosmosDBProcessor.PARTITION_KEY,MOCK_PARTITION_FIELD_NAME);
+            if(withConnectionService) {
```
```suggestion
            if (withConnectionService) {
```

jfrazee reviewed (changes requested):
@sjyang18 Thanks for addressing the previous comments. Most of what's left to do is addressing some consistency/style/whitespace conventions (I know it's tedious stuff so thanks in advance). I left comments for all it inline so it's easy to keep track of all the cases.

sjyang18 commented on nifi-nar-bundles/nifi-azure-bundle/pom.xml:
```diff
@@ -29,6 +29,8 @@
         <azure.core.version>1.5.0</azure.core.version>
         <jackson.version>2.10.3</jackson.version>
+        <azure-cosmos.version>4.2.0</azure-cosmos.version>
+        <azure-core.version>1.6.0</azure-core.version>
```
If we stay with azure core 1.5.0, this will not work with Cosmos 4.2.0, which has the better performance than Cosmos 4.1.0.

I tested out integration tests with azure core version 1.6.0, and it shows the same test result as 1.5.0. No test failures with azure-datalake processors.

sjyang18 replied:
com.azure:azure-identity (https://mvnrepository.com/artifact/com.azure/azure-identity/1.0.6) has also dependency on com.azure:azure-core, which is 1.5.0 and is upgradable to 1.9.0. Thus, it is also not breaking with 1.6.0 azure-core version.

sjyang18 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/MockTestBase.java:
```diff
@@ -0,0 +1,74 @@
+
+                // now, after enabling and setting the service, it should be valid
+                testRunner.enableControllerService(service);
+                testRunner.setProperty(AbstractAzureCosmosDBProcessor.CONNECTION_SERVICE, "connService");
```
assertValid() are called in the actual test cases. This is not the test case, but the common method used in the actual test cases.

sjyang18 commented:
I have tested and validated that azure-core 1.6.0 works with Azure data lake processor (I tested with PutAzureDatalakeStorage processor specifically).

jfrazee reviewed (changes requested):
After running through manual tests and ITs there are three small things I think we need to change:
1. A missing partition key field needs to throw an error.
2. There needs to be a check for a blank id in addition to null.
3. The controller service name should be `AzureCosmosDBClientService` to match the convention used elsewhere.
If we make these changes it should be ready to go.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/pom.xml:
```diff
@@ -27,6 +27,38 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-core</artifactId>
+            <version>${azure-core.version}</version>
```
```suggestion
            <version>${azure.core.version}</version>
```

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/pom.xml:
```diff
@@ -27,6 +27,38 @@
+            </exclusions>
         </dependency>
         <dependency>
             <groupId>com.azure</groupId>
```
This is a duplicate azure-core dependency. Need to remove.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AzureCosmosDBUtils.java:
```diff
@@ -0,0 +1,62 @@
+        .name("azure-cosmos-consistency-level")
+        .displayName("Cosmos DB Consistency Level")
+        .description("Choose from five consistency levels on the consistency spectrum. "
+            + "Refer to Cosmos DB document for their difference")
```
```suggestion
            + "Refer to Cosmos DB documentation for their difference")
```

jfrazee reviewed (changes requested):
I found a couple things while testing. The big ones are that the logic to create an id when one is missing needs to account for an "id" key that's missing entirely and there's a duplicate azure-core dependency in the services api pom file. I did some performance testing with those fixed and the performance and functionality are good.

sjyang18 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,221 @@
+            while ((record = reader.nextRecord()) != null) {
+                // Convert each Record to HashMap
+                Map<String, Object> contentMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(schema));
+                if(contentMap.containsKey("id")) {
```
I will create UUID if id does not exists.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,223 @@
+                }
+                if (!contentMap.containsKey(partitionKeyField)) {
+                    logger.error(String.format("PutAzureCosmoDBRecord failed with missing partitionKeyField (%s)", partitionKeyField));
+                    session.transfer(flowFile, REL_FAILURE);
```
```suggestion
```

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,223 @@
+
+        } catch (SchemaNotFoundException | MalformedRecordException | IOException | CosmosException e) {
+            logger.error("PutAzureCosmoDBRecord failed with error: {}", new Object[]{e.getMessage()}, e);
+            session.transfer(flowFile, REL_FAILURE);
```
```suggestion
```

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java:
```diff
@@ -0,0 +1,223 @@
+            if (!error) {
+                session.getProvenanceReporter().send(flowFile, getURI(context));
+                session.transfer(flowFile, REL_SUCCESS);
+            }
```
```suggestion
            } else {
                session.transfer(flowFile, REL_FAILURE);
            }
```

jfrazee reviewed (changes requested):
@sjyang18 The `session.transfer()` should be moved outside the scope of the `session.read()` for the failure cases for the partition key failure handling.

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/AzureCosmosDBUtils.java:
```diff
@@ -0,0 +1,62 @@
+            + " Note this host URL is for Cosmos DB with CORE SQL API"
+            + " from Azure Portal (Overview->URI)")
+        .required(false)
+        .addValidator(StandardValidators.URI_VALIDATOR)
```
```suggestion
        .addValidator(StandardValidators.URI_VALIDATOR)
        .sensitive(true)
```

MikeThomsen commented:
For Cosmos graph support, take a look at the nifi-graph-bundle folder under nifi-nar-modules and play around with the gremlin support there. Would be interested to know how well it works (or doesn't) with the Cosmos gremlin API.
